### PR TITLE
Compiler/View API documentation

### PR DESCRIPTION
**Pull request by IgorMinar**

### CODE REVIEW

File: modules/angular2/src/core/compiler/compiler.ts [hunk 1]
```diff
@@ -127,8 +132,10 @@ export class Compiler {
     return PipeBinding.createFromType(typeOrBinding, meta);
   }
 
-  // Create a hostView as if the compiler encountered <hostcmp></hostcmp>.
-  // Used for bootstrapping.
+  /**
+   * Compiles an {@link EntryPointComponent entry-point Component} and returns a promise for a
+   * {@link ProtoViewRef} that can be passed into {@link AppViewManager
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/compiler.ts]: Missing `}.` at the end.

File: modules/angular2/src/core/compiler/element_ref.ts [hunk 2]
```diff
@@ -48,15 +61,22 @@ export class ElementRef implements RenderElementRef {
...
-   * NOTE: This method will return null in the webworker scenario!
+   * <div class="callout is-critical">
+   *   <header>Use with caution</header>
+   *   <p>
+   *    Use this api as the last resort when direct access to DOM is needed. Use templating and
+   *    data-binding provided by Angular instead.
+   *   </p>
+   *   <p>
+   *    Relying on direct DOM access creates tight coupling between your application and rendering
+   *    layers which will make it impossible to separate the two and deploy your application in into a
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/element_ref.ts]: _in into_  -->  _..._

File: modules/angular2/src/core/compiler/template_ref.ts [hunk 3]
```diff
@@ -3,17 +3,26 @@ import {ElementRef} from './element_ref';
...
+ * Use {@link ViewContainerRef#createEmbeddedView} method to instantiate an Embedded View and attach
+ * it to a View Container.
+ *
+ * <!-- TODO: how to get hold of it? -->
  */
 export class TemplateRef {
+
   /**
-   * The location of the template
+   * The location in the View where the Embedded View logically belong to.
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/template_ref.ts]: _belong_  -->  _belongs_

File: modules/angular2/src/core/compiler/view_container_ref.ts [hunk 4]
```diff
@@ -10,17 +10,43 @@ import {TemplateRef} from './template_ref';
...
 
 /**
- * A location where {@link ViewRef}s can be attached.
+ * Represents a container where one or more Views can be attached.
  *
- * A `ViewContainerRef` represents a location in a {@link ViewRef} where other child
- * {@link ViewRef}s can be inserted. Adding and removing views is the only way of structurally
- * changing the rendered DOM of the application.
+ * The container can contain two kinds of Views. Host Views, created by instantiating a
+ * {@link Component} via {@link #createHostView}s, and Embedded Views, created by instantiating an
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/view_container_ref.ts]: _#createHostView}s_  -->  _#createHostView}_

File: modules/angular2/src/core/compiler/view_container_ref.ts [hunk 5]
```diff
@@ -10,17 +10,43 @@ import {TemplateRef} from './template_ref';
...
+ * {@link TemplateRef Embedded Template} via {@link #createEmbeddedView}).
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/view_container_ref.ts]: _)._  -->  _._

File: modules/angular2/src/core/compiler/view_container_ref.ts [hunk 6]
```diff
@@ -10,17 +10,43 @@ import {TemplateRef} from './template_ref';
...
+ *
+ * The location of the View Container within the containing View is specified by the Anchor
+ * `element`. Each View Container can have only one Anchor Element and each Anchor Element can only
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/view_container_ref.ts]: _\`element\`_  -->  _Element_ (?)

File: modules/angular2/src/core/compiler/view_container_ref.ts [hunk 7]
```diff
@@ -37,80 +63,93 @@ export class ViewContainerRef {
...
-    if (atIndex == -1) atIndex = this.length;
-    return this.viewManager.attachViewInContainer(this.element, atIndex, viewRef);
+  insert(viewRef: ViewRef, index: number = -1): ViewRef {
+    if (index == -1) index = this.length;
+    return this.viewManager.attachViewInContainer(this.element, index, viewRef);
   }
 
   /**
-   * Return the index of already inserted {@link ViewRef}.
+   * Returns the index of the View, specified via {@link ViewRef}, within the current container.
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/view_container_ref.ts]: _{@link ViewRef}_  -->  _{@link ViewRef `viewRef`}_ ?

File: modules/angular2/src/core/compiler/view_ref.ts [hunk 8]
```diff
@@ -12,21 +12,27 @@ export function internalProtoView(protoViewRef: ProtoViewRef): viewModule.AppPro
   return isPresent(protoViewRef) ? protoViewRef._protoView : null;
 }
 
+
+/**
+ * Represents the View of a {@link Component} that was compiled via {@link Compiler#compileInHost
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/view_ref.ts]: Missing `}.` at the end.

File: modules/angular2/src/core/compiler/view_ref.ts [hunk 9]
```diff
@@ -91,13 +97,14 @@ export class ViewRef implements HostViewRef {
...
 /**
- * A reference to an Angular ProtoView.
+ * Represents Angular's ProtoView.
+ *
+ * A ProtoView is a prototypical View that is the result of Template compilation and enables
+ * Angular to efficiently create an instance of a View based on the compiled Template.
  *
- * A ProtoView is a reference to a template for easy creation of views.
- * (See {@link AppViewManager#createViewInContainer `AppViewManager#createViewInContainer`} and
+ * A ProtoView is created {@link AppViewManager#createViewInContainer} and
```
Comment (gkalpak) [inline modules/angular2/src/core/compiler/view_ref.ts]: _created_  -->  _created with/via/using_

_{@link AppViewManager#createViewInContainer}_  -->  _{@link AppViewManager#createViewInContainer \`AppViewManager#createViewInContainer\`}_
(for consistency with `AppViewManager#createRootHostView`)

File: modules/angular2/src/core/compiler/element_ref.ts (hunk 2 above)
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/element_ref.ts]: thanks

File: modules/angular2/src/core/compiler/view_container_ref.ts (hunk 4 above)
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/view_container_ref.ts]: thanks

File: modules/angular2/src/core/compiler/view_container_ref.ts (hunk 5 above)
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/view_container_ref.ts]: thanks

File: modules/angular2/src/core/compiler/view_container_ref.ts (hunk 6 above)
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/view_container_ref.ts]: the property is going to be renamed to anchorElement which will make the docs less awkward

**Comment by IgorMinar** [comment]
@tbosch can you please review. This PR doesn't contain most of the property/method/local var renames that we discussed.

**Comment by IgorMinar** [comment]
and I need to rebase it on master, but you can review the docs even before that

File: modules/angular2/src/core/compiler/dynamic_component_loader.ts [hunk 10]
```diff
@@ -43,39 +60,58 @@ export class ComponentRef {
...
+
+  /**
+   * @private
+   */
   constructor(private _compiler: Compiler, private _viewManager: AppViewManager) {}
 
   /**
-   * Loads a root component that is placed at the first element that matches the component's
-   * selector.
+   * Creates an instance of a Component and attaches it to the first element in the global View
```
Comment (tbosch) [inline modules/angular2/src/core/compiler/dynamic_component_loader.ts]: "View in "global view" is not an Angular view. Maybe find a better word?

File: modules/angular2/src/core/compiler/dynamic_component_loader.ts [hunk 11]
```diff
@@ -43,39 +60,58 @@ export class ComponentRef {
...
+   * (usually DOM Document) that matches the component's selector.
+   *
+   * <!-- TODO(i): document parameters and return value -->
    *
    * - `typeOrBinding` `Type` \ {@link Binding} - representing the component to load.
```
Comment (tbosch) [inline modules/angular2/src/core/compiler/dynamic_component_loader.ts]: This has already changed to just be a `Type` in master. Same for all the other methods and `Compiler.compileInHost`.

File: modules/angular2/src/core/compiler/dynamic_component_loader.ts [hunk 12]
```diff
@@ -43,39 +60,58 @@ export class ComponentRef {
...
    * - `overrideSelector` (optional) selector to load the component at (or use
```
Comment (tbosch) [inline modules/angular2/src/core/compiler/dynamic_component_loader.ts]: maybe better: css selector used to locate the element to which the component should be attached

File: modules/angular2/src/core/compiler/dynamic_component_loader.ts [hunk 13]
```diff
@@ -141,10 +177,12 @@ export class DynamicComponentLoader {
   }
 
   /**
-   * Loads a component into the component view of the provided ElementRef next to the element
-   * with the given name.
+   * Creates an instance of a Component and attaches to a Component View of a Component instance.
```
Comment (tbosch) [inline modules/angular2/src/core/compiler/dynamic_component_loader.ts]: Creates an instance of a Component and attaches it next to a location inside of the Component View of another Component instance

File: modules/angular2/src/core/compiler/element_ref.ts [hunk 14]
```diff
@@ -3,20 +3,28 @@ import {ViewRef} from './view_ref';
 import {RenderViewRef, RenderElementRef, Renderer} from 'angular2/src/core/render/api';
 
 /**
- * An opaque reference to the underlying element.
+ * Represents a location in a View that has an injection, change-detection and render contexts
```
Comment (tbosch) [inline modules/angular2/src/core/compiler/element_ref.ts]: singular `context`

File: modules/angular2/src/core/compiler/template_ref.ts [hunk 15]
```diff
@@ -3,17 +3,35 @@ import {ElementRef} from './element_ref';
 import * as viewModule from './view';
 
 /**
- * Reference to a template within a component.
+ * Represents an Embedded Template that can be used to instantiate Embedded Views.
  *
- * Represents an opaque reference to the underlying template that can
- * be instantiated using the {@link ViewContainerRef}.
+ * You can access a TemplateRef, via a directive placed on a `<template>` element (or directive
```
Comment (tbosch) [inline modules/angular2/src/core/compiler/template_ref.ts]: `TemplateRef` can also be queries via a view query.

File: modules/angular2/src/core/compiler/template_ref.ts [hunk 16]
```diff
@@ -23,10 +41,16 @@ export class TemplateRef {
         .nestedProtoView;
   }
 
+  /**
+   * @private
+   *
+   * Reference to the ProtoView, created by compiling the an Embedded Template, using which Embedded
```
Comment (tbosch) [inline modules/angular2/src/core/compiler/template_ref.ts]: "the an"

File: modules/angular2/src/core/compiler/dynamic_component_loader.ts (hunk 11 above)
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/dynamic_component_loader.ts]: yes. I need to rebase

File: modules/angular2/src/core/compiler/element_ref.ts (hunk 14 above)
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/element_ref.ts]: done

File: modules/angular2/src/core/compiler/template_ref.ts (hunk 15 above)
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/template_ref.ts]: done

File: modules/angular2/src/core/compiler/template_ref.ts (hunk 16 above)
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/template_ref.ts]: done

File: modules/angular2/src/core/compiler/dynamic_component_loader.ts (hunk 13 above)
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/dynamic_component_loader.ts]: done

File: modules/angular2/src/core/compiler/dynamic_component_loader.ts (hunk 12 above)
Comment (IgorMinar) [inline modules/angular2/src/core/compiler/dynamic_component_loader.ts]: done

**Pull request closed without merging**